tests/gradients/test_parameter_shift_cv.py

```diff
@@ -0,0 +1,1105 @@
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Rotation(1.0, wires=[0])
+            qml.expval(qml.NumberOperator(0))
+
+        tape.trainable_params = {0}
+
+        with pytest.warns(UserWarning, match="does not support the PolyXP observable"):
+            qml.gradients.param_shift_cv(tape, dev)
+
+        spy_second_order.assert_not_called()
+
+    def test_no_poly_xp_support_variance(self, mocker, monkeypatch, caplog):
+        """Test that if a device does not support PolyXP
+        and the variance parameter-shift rule is required,
+        we fallback to finite differences."""
+        spy = mocker.spy(qml.gradients.parameter_shift_cv, "var_param_shift")
+        dev = qml.device("default.gaussian", wires=1)
+
+        monkeypatch.delitem(dev._observable_map, "PolyXP")
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Rotation(1.0, wires=[0])
+            qml.var(qml.X(0))
+
+        tape.trainable_params = {0}
+
+        with pytest.warns(UserWarning, match="does not support the PolyXP observable"):
+            qml.gradients.param_shift_cv(tape, dev)
+
+        spy.assert_not_called()
+
+    def test_independent_parameters_analytic(self):
+        """Test the case where expectation values are independent of some parameters. For those
+        parameters, the gradient should be evaluated to zero without executing the device."""
+        dev = qml.device("default.gaussian", wires=2)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Displacement(1, 0, wires=[1])
+            qml.Displacement(1, 0, wires=[0])
+            qml.expval(qml.X(0))
+
+        tape.trainable_params = {0, 2}
+        tapes, fn = qml.gradients.param_shift_cv(tape, dev)
+
+        # We should only be executing the device to differentiate 1 parameter
+        # (first order, so 2 executions)
+        assert len(tapes) == 2
+
+        res = fn(dev.batch_execute(tapes))
+        assert np.allclose(res, [0, 2])
+
+        tape.trainable_params = {0, 2}
+        tapes, fn = qml.gradients.param_shift_cv(tape, dev, force_order2=True)
+
+        # We should only be executing the device to differentiate 1 parameter
+        # (second order, so 0 executions)
+        assert len(tapes) == 0
+
+        res = fn(dev.batch_execute(tapes))
+        assert np.allclose(res, [0, 2])
+
+    def test_all_independent(self):
+        """Test the case where expectation values are independent of all parameters."""
+        dev = qml.device("default.gaussian", wires=2)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Displacement(1, 0, wires=[1])
+            qml.Displacement(1, 0, wires=[1])
+            qml.expval(qml.X(0))
+
+        tape.trainable_params = {0, 2}
+        tapes, fn = qml.gradients.param_shift_cv(tape, dev)
+        assert len(tapes) == 0
+
+        grad = fn(dev.batch_execute(tapes))
+        assert np.allclose(grad, [0, 0])
+
+
+class TestExpectationQuantumGradients:

```

Comment:
Copy and paste of: https://github.com/PennyLaneAI/pennylane/blob/ab71001bc55c2945a0945f42131c45d77da3512e/tests/tape/test_cv_param_shift.py#L356